src/cargo/sources/registry/remote.rs

```diff
@@ -35,6 +36,8 @@ pub struct RemoteRegistry<'cfg> {
     repo: LazyCell<git2::Repository>,
     head: Cell<Option<git2::Oid>>,
     current_sha: Cell<Option<InternedString>>,
+    needs_update: Cell<bool>,
+    updated: bool,
```

Comment:
Generally `RemoteRegistry` is making my head spin. How do `needs_update` and `updated` interact? What code is new vs. what has just moved?

Do we have tests that the index is not fetched to much? I.E. At most once per invocation to `cargo` and none if the lockfile works.

I will look again when I am fresh and it may be clearer.

Reply:
Most of the code just moved. It was previously in `update`, and had to move to outside the trait implementation into `update_index_worker`. 

There are some functional changes to `load`. Previously, it would error out if a crate was unavailable, then layers above would catch the error, `update`, and try again. Now `load` keeps track itself. It will only error out if an `update` has happened.

`needs_update` is associated with `invalidate_cache`. It means that someone as requested that this source must fetch the index for the current cargo session.

`updated` is to prevent fetching the index more than once.

Yes, there are tests covering those scenarios. I hit failures when I was implementing this :).

Reply:
That makes sense now.